src/sys/windows/selector.rs

```diff
@@ -8,7 +8,8 @@ use crate::{Interests, Token};
 use miow::iocp::{CompletionPort, CompletionStatus};
 use miow::Overlapped;
 use std::collections::VecDeque;
-use std::mem::size_of;
+use std::marker::PhantomPinned;
+use std::mem::{forget, size_of, transmute_copy};
```

Comment:
Is there any way we can not use `transmute_copy`?

Reply:
I don't agree that `transmute_copy` is specifically "very unsafe", it is in fact safer than performing a raw pointer cast.
E.g. it could be done as follows:

```
fn update(&mut self, self_arc: &Pin<Arc<Mutex<SockState>>>) -> io::Result<()> {
  // <snip>
  let overlapped = unsafe { *(self_arc as *const _ as *const PVOID) };
```

... but that does not even validate that `Arc<T>` and `void*` have the same size.
So, unless there are other reasons to switch, I'm sticking with transmute_copy.

Reply:
I forgot to mention:

The more readable solution would have been to use `Arc::into_raw(self_arc.clone())` and `Arc::from_raw()`.
Unfortunately that doesn't work in this case because we need to ensure the reference count is increased only if the overlapped AFD_POLL operation starts successfully, but we need the pointer to try starting it. 

🐔/🥚, so to speak.

Reply:
> ... but that does not even validate that `Arc<T>` and `void*` have the same size.

Neither does `transmute_copy`, in the docs it states "It is not a compile-time error if `T` and `U` have different sizes, but it is highly encouraged to only invoke this function where `T` and `U` have the same size. This function triggers undefined behavior if `U` is larger than `T`." (https://doc.rust-lang.org/std/mem/fn.transmute_copy.html)

Can't we increase the ref counter using `Arc::into_raw(self_arc.clone())` and then if we fail reduce it again? And if we succeed just continue?

Reply:
> Neither does transmute_copy, in the docs it states "It is not a compile-time error if T and U have different sizes, but it is highly encouraged to only invoke this function where T and U have the same size. This function triggers undefined behavior if U is larger than T." (https://doc.rust-lang.org/std/mem/fn.transmute_copy.html)

I didn't realize transmute_copy behaves different than transmute() in this regard.

> Can't we increase the ref counter using Arc::into_raw(self_arc.clone()) and then if we fail reduce it again? And if we succeed just continue?

It is possible, but is it better?

I have no interest in churning this PR, I have other things to do.
Feel free to push to this branch and land it once it you think it is acceptable.

Reply:
@Thomasdezeeuw here's another idea, why not ensuring token is unique for every registered socket and use the token as overlapped instead of playing with pointers and reference counting. Maybe I'm missing something but, Pin and transmute_copy doesn't seem to bring any benefit over the previous solution with overlapped_wrapepr. 

If token is uniqe, we could easily id a socket when you get a completion callback, no more playing with reference, and unsafe pointers, also it is going to work nicely with deleting sockets on selector drop (the fix for mem leaks).

Reply:
In idiomatic rust, it is far preferred to do the raw pointer deref / cast than using `transmute`. Generally, one should avoid all transmute unless there is no other way to achieve the goal.